exec/java-exec/src/main/java/org/apache/drill/exec/store/log/LogRecordReader.java

```diff
@@ -0,0 +1,750 @@
+      default:
+        throw UserException
+            .validationError()
+            .message("Undefined column types")
+            .addContext("Position", patternIndex)
+            .addContext("Field name", name)
+            .addContext("Type", typeName)
+            .build(logger);
+    }
+  }
+
+  private void defineVectors() {
+    for (int i = 0; i < columns.length; i++) {
+      try {
+        columns[i].define(outputMutator);
+      } catch (SchemaChangeException e) {
+        throw UserException
+            .systemError(e)
+            .message("Vector creation failed")
+            .build(logger);
+      }
+    }
+  }
+
+  @Override
+  public int next() {
+    rowIndex = 0;
+    while (nextLine()) {
+    }
+    return rowIndex;
+  }
+
+  private boolean nextLine() {
+    String line;
+    try {
+      line = reader.readLine();
+    } catch (IOException e) {
+      throw UserException
+          .dataReadError(e)
+          .message("Error reading file:")
+          .addContext("File", fileWork.getPath())
+          .build(logger);
+    }
+
+    if (line == null) {
+      return false;
+    }
+    Matcher lineMatcher = pattern.matcher(line);
+    if (lineMatcher.matches()) {
+      loadVectors(lineMatcher);
+      return rowIndex < BATCH_SIZE;
+    }
+
+    errorCount++;
+    if (errorCount < formatConfig.getMaxErrors()) {
+      logger.warn("Unmatached line: {}", line);
+    } else if (errorCount > formatConfig.getMaxErrors()) {
+      throw UserException.parseError()
+          .message("Too many errors.  Max error threshold exceeded.")
+          .addContext("Line", line)
+          .addContext("Line number", rowIndex)
+          .build(logger);
+    }
+    //If the user asked for the unmatched columns display them
+    if (unmatchedRows) {
+      //If the user asked for the unmatched columns AND other columns
+      if (columns.length > 1) {
+        columns[unmatchedColumnIndex].load(rowIndex, line);
+        rowIndex++;
+        return rowIndex < BATCH_SIZE;
+      } else {
+        //If the user ONLY asked for the unmatched columns
+        columns[unmatchedColumnIndex].load(unmatchedRowIndex, line);
+        unmatchedRowIndex++;
+        rowIndex = unmatchedRowIndex;
+        return unmatchedRowIndex < BATCH_SIZE;
+      }
+    }
+
+    return true;
+  }
+
+  private void loadVectors(Matcher m) {
+    String value = null;
+    /*if( unmatchedRows && columns.length == 1 ){
+      return;
+    }*/
+
+    for (int i = 0; i < columns.length; i++) {
+      //Skip the unmatched rows column
+      if( columns[i].name.equals("_unmatched_rows")){
+        continue;
+      }
+
+      if (columns[i].index >= 0) {
+        //Get the value of the regex group
+        value = m.group(columns[i].index + 1);
+
+        //If the value is not null, assign it to the column
+        if (value != null) {
+          columns[i].load(rowIndex, value);
+        }
+      } else if (columns[i].name.equals("_raw")) {
```

Comment:
Seems awkward.

During setup, there should be a choice: either pick out all columns, or just pick out the raw row. This means that we either get the raw row, or we get the columns.

Otherwise, handle it by updating the column class methods. Perhaps those columns that implement `isSpecial()` are given the entire row. Just making something up:

```
  if (m.isSpecial()) {
    ((RowColumn) m).loadRow(rowIndex, m.group(0), hasMatch);
  }
```

That is, how can we use method dispatch to a subclass (which is relatively fast) to substitute for if-statements with string comparisons (which is slow)?

Reply:
Hi @paul-rogers 
Not necessarily.   There are cases where you might want the `_raw` row AND the fields.  Debugging the regex for example.  You might want to compare the raw data with what Drill is extracting.  

I could think of other situations where you have a complex log, but are only extracting a few select fields by default.